tket/src/WeightSubgrMono/GraphTheoretic/DomainInitialiser.cpp

```diff
@@ -0,0 +1,153 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/GraphTheoretic/DomainInitialiser.hpp"
+
+#include <algorithm>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/Common/SetIntersection.hpp"
+#include "WeightSubgrMono/GraphTheoretic/FilterUtils.hpp"
+#include "WeightSubgrMono/GraphTheoretic/NearNeighboursData.hpp"
+#include "WeightSubgrMono/GraphTheoretic/NeighboursData.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+bool DomainInitialiser::full_initialisation(
+    InitialDomains& initial_domains,
+    const NeighboursData& pattern_neighbours_data,
+    NearNeighboursData& pattern_near_neighbours_data,
+    const NeighboursData& target_neighbours_data,
+    NearNeighboursData& target_near_neighbours_data, unsigned max_path_length) {
+  return degree_sequence_initialisation(
+             initial_domains, pattern_neighbours_data,
+             target_neighbours_data) &&
+
+         distance_counts_reduction(
+             initial_domains, pattern_near_neighbours_data,
+             target_near_neighbours_data, max_path_length);
+}
+
+bool DomainInitialiser::degree_sequence_initialisation(
+    InitialDomains& initial_domains,
+    const NeighboursData& pattern_neighbours_data,
+    const NeighboursData& target_neighbours_data) {
+  initial_domains.resize(
+      pattern_neighbours_data.get_number_of_nonisolated_vertices());
+  for (auto& domain : initial_domains) {
+    domain.clear();
+  }
+
+  // Get the target degree sequences.
+  const auto number_of_tv =
+      target_neighbours_data.get_number_of_nonisolated_vertices();
+
+  // FIRST: the target vertex;
+  // SECOND: its degree sequence.
+  typedef std::pair<VertexWSM, std::vector<std::size_t>> DegSeqData;
+  std::vector<DegSeqData> target_degree_sequences;
+
+  target_degree_sequences.reserve(number_of_tv);
+  for (unsigned tv = 0; tv < number_of_tv; ++tv) {
+    target_degree_sequences.emplace_back(
+        tv, target_neighbours_data.get_sorted_degree_sequence_expensive(tv));
+  }
+
+  // Sort by decreasing sequence length.
+  std::sort(
+      target_degree_sequences.begin(), target_degree_sequences.end(),
+      [](const DegSeqData& lhs, const DegSeqData& rhs) -> bool {
+        return lhs.second.size() > rhs.second.size() ||
+               // This is not crucial, it just ensures exact behaviour for
+               // nonstable sorts:
+               (lhs.second.size() == rhs.second.size() &&
+                lhs.first < rhs.first);
+      });
+
+  // TODO: get TV data lazily (maybe some will be unused).
```

Comment:
Can TODO be removed?

Reply:
Only by doing it! Of course it CAN be done, there's just never enough time!

Reply:
Ha, yeah I guess. Might be worth adding a basic task to the Backlog in the hope it might get prioritised at some point

Reply:
Yes, made the JIRA: https://cqc.atlassian.net/browse/TKET-2233